File: src/main/java/net/frozenblock/configurableeverything/mod_compat/ConfigurableEverythingIntegrations.java
package net.frozenblock.configurableeverything.mod_compat;

import java.util.function.Supplier;
import net.frozenblock.configurableeverything.util.ConfigurableEverythingSharedConstants;
import net.frozenblock.lib.integration.api.ModIntegration;
import net.frozenblock.lib.integration.api.ModIntegrationSupplier;
import net.frozenblock.lib.integration.api.ModIntegrations;

public final class ConfigurableEverythingIntegrations {

	public static final ModIntegration FROZENLIB_INTEGRATION = registerAndGet(FrozenLibIntegration::new, "frozenlib");

	private ConfigurableEverythingIntegrations() {
		throw new UnsupportedOperationException("ConfigurableEverythingIntegrations contains only static declarations.");
	}

	public static void init() {
	}

	public static ModIntegrationSupplier<? extends ModIntegration> register(Supplier<? extends ModIntegration> integration, String modID) {
		return ModIntegrations.register(integration, ConfigurableEverythingSharedConstants.MOD_ID, modID);
	}

	public static <T extends ModIntegration> ModIntegrationSupplier<T> register(Supplier<T> integration, Supplier<T> unloadedIntegration, String modID) {
		return ModIntegrations.register(integration, unloadedIntegration, ConfigurableEverythingSharedConstants.MOD_ID, modID);
	}

	public static <T extends ModIntegration> ModIntegration registerAndGet(Supplier<T> integration, String modID) {
		return ModIntegrations.register(integration, ConfigurableEverythingSharedConstants.MOD_ID, modID).getIntegration();
	}
}

File: src/main/java/net/frozenblock/configurableeverything/mod_compat/ConfigurableEverythingIntegrations.kt
package net.frozenblock.configurableeverything.mod_compat

import net.frozenblock.configurableeverything.mod_compat.FrozenLibIntegration
import net.frozenblock.configurableeverything.util.MOD_ID
import net.frozenblock.lib.integration.api.ModIntegration
import net.frozenblock.lib.integration.api.ModIntegrationSupplier
import net.frozenblock.lib.integration.api.ModIntegrations
import java.util.function.Supplier

internal object ConfigurableEverythingIntegrations {

    @JvmField
    internal val FROZENLIB_INTEGRATION: ModIntegration = registerAndGet({ FrozenLibIntegration }, "frozenlib")

    inline fun register(
        crossinline integration: () -> ModIntegration?,
        modID: String?
    ): ModIntegrationSupplier<out ModIntegration> {
        return ModIntegrations.register(Supplier { integration() }, MOD_ID, modID)
    }

    inline fun <T : ModIntegration?> register(
        crossinline integration: () -> T?,
        crossinline unloadedIntegration: () -> T?,
        modID: String?
    ): ModIntegrationSupplier<T> {
        return ModIntegrations.register(Supplier { integration() }, Supplier { unloadedIntegration() }, MOD_ID, modID)
    }

    fun <T : ModIntegration?> registerAndGet(integration: () -> T?, modID: String?): ModIntegration {
        return ModIntegrations.register(Supplier { integration() }, MOD_ID, modID).integration
    }
}
